abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+
+#### `proxy_set_metric_value`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i64 (uint64_t) value`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set the value of the metric (`metric_id`) to the value (`value`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the metric (`metric_id`) doesn't exist.
+
+Returns `InvalidOperation` when the `metric_type` doesn't support setting
+values.
+
+#### `proxy_increment_metric_value`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i64 (int64_t) offset`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Increment/decrement value of the metric (`metric_id`) by offset (`offset`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the metric (`metric_id`) doesn't exist.
+
+Returns `InvalidOperation` when `metric_type` doesn't support incrementing
+and/or decrementing values.
+
+#### `proxy_delete_metric`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Deletes previously created metric (`metric_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the metric (`metric_id`) doesn't exist.
+
+
+## HTTP callouts
+
+### Functions exposed by the host environment
+
+#### `proxy_dispatch_http_call`
+
+* params:
+  - `i32 (const char*) upstream_name_data`
+  - `i32 (size_t) upstream_name_size`
+  - `i32 (iovec_array_t[]) headers_map_data`
+  - `i32 (size_t) headers_map_size`
+  - `i32 (iovec_array_t) body_data`
+  - `i32 (size_t) body_size`
+  - `i32 (iovec_array_t[]) trailers_map_data`
+  - `i32 (size_t) trailers_map_size`
+  - `i32 (uint32_t) timeout_milliseconds`
+  - `i32 (uint32_t*) return_callout_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Dispatches a HTTP call to upstream (`upstream_name_data`, `upstream_name_size`).
+
+Once the response is returned to the host, `proxy_on_http_call_response` will be
+called with an unique call identifier (`return_callout_id`).
+
+Returns `Ok` on success.
+
+Returns `NotFound` for unknown upstream.
+
+Returns `NotAllowed` when requests to given upstream are not allowed.
+
+Returns `InvalidMemoryAccess` when `upstream_name_data`, `upstream_name_size`,
+`headers_map_data`, `headers_map_size`, `body_data`, `body_size`,
+`trailers_map_data`, `trailers_map_size` and/or `return_callout_id` point to
+invalid memory address.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_http_call_response`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+  - `i32 (size_t) num_headers`
+  - `i32 (size_t) body_size`
+  - `i32 (size_t) num_trailers`
+* returns:
+  - none
+
+Called when the complete response to the HTTP call (`callout_id`) is received.
+
+Headers (`num_headers` entries) and trailers (`num_trailers` entries) can be
+retrieved using `proxy_get_map_values`.
+
+Response body (of size `body_size`) can be retrieved using `proxy_get_buffer`.
+
+
+## gRPC callouts
+
+### Functions exposed by the host environment
+
+#### `proxy_dispatch_grpc_call`
```

Comment:
Why is there a difference in the ABI between a call and a stream? Why can't we just treat everything like a stream and let higher level code deal with the distinction?

Reply:
The only difference I think is in terms of ideas like timeout, which the SDK can potentially handle itself.